xformers/components/attention/csrc/cuda/attention.cu

```diff
@@ -0,0 +1,592 @@
+  out[0].x *= x1;
+  out[0].y *= x1;
+  out[0].z *= x1;
+  out[0].w *= x1;
+}
+
+template <typename scalar_t>
+__device__ __forceinline__ void iMul(scalar_t x1, float2* out) {
+  out[0].x *= x1;
+  out[0].y *= x1;
+}
+
+template <typename scalar_t>
+__device__ __forceinline__ void iMul(scalar_t x1, float* out) {
+  out[0] *= x1;
+}
+
+template <typename scalar_t>
+__device__ __forceinline__ void iDiv(scalar_t x1, float4* out) {
+  out[0].x /= x1;
+  out[0].y /= x1;
+  out[0].z /= x1;
+  out[0].w /= x1;
+}
+
+template <typename scalar_t>
+__device__ __forceinline__ void iDiv(scalar_t x1, float2* out) {
+  out[0].x /= x1;
+  out[0].y /= x1;
+}
+
+template <typename scalar_t>
+__device__ __forceinline__ void iDiv(scalar_t x1, float* out) {
+  out[0] /= x1;
+}
+
+template <typename scalar_t, int WARP_SIZE>
+__device__ __forceinline__ scalar_t warpSum(scalar_t val) {
+  for (int stride = WARP_SIZE / 2; stride > 0; stride >>= 1) {
+    val += __shfl_xor_sync(0xffffffff, val, stride, WARP_SIZE);
+  }
+  return val;
+}
+
+template <typename scalar_t, int WARP_SIZE>
+__device__ __forceinline__ float2 warpSum(float2 val) {
+  for (int stride = WARP_SIZE / 2; stride > 0; stride >>= 1) {
+    val.x += __shfl_xor_sync(0xffffffff, val.x, stride, WARP_SIZE);
+    val.y += __shfl_xor_sync(0xffffffff, val.y, stride, WARP_SIZE);
+  }
+  return val;
+}
+
+template <typename scalar_t, int WARP_SIZE>
+__device__ __forceinline__ float4 warpSum(float4 val) {
+  for (int stride = WARP_SIZE / 2; stride > 0; stride >>= 1) {
+    val.x += __shfl_xor_sync(0xffffffff, val.x, stride, WARP_SIZE);
+    val.y += __shfl_xor_sync(0xffffffff, val.y, stride, WARP_SIZE);
+    val.z += __shfl_xor_sync(0xffffffff, val.z, stride, WARP_SIZE);
+    val.w += __shfl_xor_sync(0xffffffff, val.w, stride, WARP_SIZE);
+  }
+  return val;
+}
+
+template <typename scalar_t, int WARP_SIZE>
+__device__ __forceinline__ scalar_t warpMax(scalar_t val) {
+  for (int stride = WARP_SIZE / 2; stride > 0; stride >>= 1) {
+    scalar_t tmp = __shfl_xor_sync(0xffffffff, val, stride, WARP_SIZE);
+    val = tmp > val ? tmp : val;
+  }
+  return val;
+}
+
+template <typename scalar_t, typename vec_t, int kBlockSizeK, int kBlockSizeQ>
+__device__ void compute_dot(
+    vec_t* queries[kBlockSizeQ],
+    vec_t* keys,
+    scalar_t out[kBlockSizeQ][kBlockSizeK],
+    int64_t K) {
+  constexpr int kVecSize = sizeof(vec_t) / sizeof(scalar_t);
+  vec_t q_i[kBlockSizeQ];
+  for (int64_t k = 0; k < K / kVecSize; k += 1) {
+#pragma unroll
+    for (int64_t q_item_idx = 0; q_item_idx < kBlockSizeQ; q_item_idx++) {
+      q_i[q_item_idx] = __ldg(queries[q_item_idx] + k);
+    }
+#pragma unroll
+    for (int64_t k_item_idx = 0; k_item_idx < kBlockSizeK; k_item_idx++) {
+      vec_t k_i = keys[k + K / kVecSize * k_item_idx];
+#pragma unroll
+      for (int64_t q_item_idx = 0; q_item_idx < kBlockSizeQ; q_item_idx++) {
+        sputnik::VectorCompute<vec_t>::Dot(
```

Comment:
@tgale96 pulling you in, just in case there's something here that we can do better (black box for me, does VectorCompute imply tensor cores for instance or normal float ops ?)

Reply:
to the best of my knowledge we don't use any tensor cores explicitly here.

This was a refactoring that I did in https://github.com/facebookresearch/xformers/pull/267/commits/cec04e917ce53efa35d1486bfe60a05061ef6b17 to simplify a few things, and it turned out that sputnik already implemented some of the things I needed, so I just took those functions

Reply:
alrighty, I think that A100 may show a little worse perf (ping @suchenzang if you're interested in giving this a go ?), but in fp32 it will be fine perf wise up to V100 included (personal guess), not a big problem. The rest of the model can still be fp16 with the torch amp guards, so all in all I think that it's probably still very fast and useful